bookkeeper-common/src/main/java/org/apache/bookkeeper/util/OrderedScheduler.java

```diff
@@ -0,0 +1,346 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package org.apache.bookkeeper.util;
+
+import com.google.common.base.Objects;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.bookkeeper.common.concurrent.FutureUtils;
+
+/**
+ * Ordered Scheduler. It is thread pool based {@link ScheduledExecutorService}, additionally providing the ability to
+ * execute/schedule tasks by <code>key</code>. Hence the tasks submitted by same <i>key</i>
+ * will be executed in order.
+ *
+ * <p>The scheduler is comprised of multiple {@link ScheduledExecutorService}s. Each {@link ScheduledExecutorService}
+ * is a single thread executor. Normal task submissions will be submitted to executors in a random manner to guarantee
+ * load balancing. Keyed task submissions (e.g {@link OrderedScheduler#submit(Object, Runnable)} will be submitted to a
+ * dedicated executor based on the hash value of submit <i>key</i>.
+ */
+public class OrderedScheduler implements ScheduledExecutorService {
```

Comment:
Hey @eolivelli with the following change we introduced scheduling methods/features to OrderedSafeExecutor, do we still need this new class / Executor? cannt we just enhance existing OrderedSafeExecutor if something is missing?

https://github.com/apache/bookkeeper/commit/c813b3d3298586ded02032a57a99d5fc6c974581#diff-9f344d1e7e7a8dfdf5e677ba83f4943f

Reply:
@reddycharan I have ported the full package from DistributedLog, I think the soon these classes will be dropped from DL. I think that a good step in the future will be to move our (BK) OrderedSafeExecutor in this new bookkeeper commons module and merge the two classes.
By now it is just a "move class" from DL.
I hope it is OK for you to keep this now here
/cc @sijie

Reply:
@eolivelli  How about make class OrderedSafeExecutor extends OrderedScheduler. then all the old reference in bk will not need change, and in future dl will also survive?
Just for your reference, opened PR #528 for port retries and backoff, which also include  our previous work of port FutureUtils and this OrderedSafeExecutor related changes.

Reply:
@zhaijack This is an interesting idea for me
I would like to move forward on this patch and IMO we can do this change later, it will be fine to me as far as this two classes have the same semantics.
This patch has already become huge.

Honestly I would like to deliver this patch as soon as possible so that I can start sending PRs for BP-14

//cc @sijie  @reddycharan

Reply:
sure. it is more import for this PR to merge.

Reply:
@eolivelli sgtm.